src/Microsoft.ML.FastTree/Training/TreeLearners/LeastSquaresRegressionTreeLearner.cs

```diff
@@ -479,17 +481,92 @@ protected virtual void FindAndSetBestFeatureForLeaf(LeafSplitCandidates leafSpli
+
+            //During feature split evaluation step the accumalated gain is subtracted from gainShift and multiplied with trust and 
+            //penalty. The below equation seperates out this quantity from each feature gain so that the cummalative 
+            //feature gain is roughly the same as the gain calculated in the feature split evaluation step.
+            double gainResidual = (gainShift * trust + usePenalty) / splitInfo.CategoricalFeatureIndices.Length;
+            for (int i = 0; i < splitInfo.CategoricalFeatureIndices.Length; i++)
+            {
+                //REVIEW mzs: For categorical split points some gain values can be negative because when gain
+                //map is built we have to reverse engineer what each feature-value's gain was from the total 
+                //gain of split at that node. The reason we chose to do this way was to speed things up and 
+                //reduce memory footprint during training time. A better solution would be to keep track 
+                //of each split point's contribution during training but that is an expensive approach.
+                int featureIndex = splitInfo.CategoricalFeatureIndices[i] - featureMin;
+
+                Contracts.Assert(featureIndex >= 0);
+
+                var binStats = stats.GetBinStats(stats.GetMinBorder(featureIndex));
+                sumGTTargets += binStats.SumTargets;
+                if (HasWeights)
+                    sumGTWeights += binStats.SumWeights;
+
+                if (binStats.Count > 0)
+                {
+                    gtCount += binStats.Count;
+                    int lteCount = totalCount - gtCount;
+                    double currentShiftedGain = GetLeafSplitGain(gtCount, sumGTTargets, sumGTWeights)
+                                                +
+                                                GetLeafSplitGain(lteCount, sumTargets - sumGTTargets,
+                                                    sumWeights - sumGTWeights);
+
+                    if (EntropyCoefficient > 0)
+                    {
+                        double entropyGain = (totalCount * Math.Log(totalCount) - lteCount * Math.Log(lteCount) -
+                                              gtCount * Math.Log(gtCount));
+
+                        currentShiftedGain += EntropyCoefficient * entropyGain;
+                    }
+
+                    splitInfo.CategoricalFeatureGain[i] = (currentShiftedGain - previousShiftGain) * trust - gainResidual;
+                    previousShiftGain = currentShiftedGain;
+                }
+                else
+                    splitInfo.CategoricalFeatureGain[i] = -1 * gainResidual;
+            }
+
+#if DEBUG
+            double sumGain = splitInfo.CategoricalFeatureGain.Sum();
+            double min = Math.Min(sumGain, splitInfo.Gain);
+            double max = Math.Max(sumGain, splitInfo.Gain);
+
+            Contracts.Assert(min / max >= 0.99999);
```

Comment:
>Contracts.Assert(min / max >= 0.99999); [](start = 12, length = 39)

Due to imprecisions in floating point calculations, I am not certain this will always be true. This might be something you'd put in a *test* of one particular problem, but the idea of putting it into the actual code that runs for all cases makes me nervous. I also don't like having `#if DEBUG` unless there's a *really* great reason for it, since every block like that is a potential source of disparity between the two.

You probably found this useful during development, but I might prefer if we got rid of this.

Reply:
You are right, with many of nodes this might go below 0.99

---
In reply to: [194039690](https://github.com/dotnet/machinelearning/pull/320#discussion_r194039690) [](ancestors = 194039690)